### ISSUE TITLE
The ability to use an expression reference, which returns a retrieve, inside of another retrieve

### ISSUE THREAD

**Issue by jackcmeyer** [cqframework/clinical_quality_language#116]
The following CQL passes validation: 

```
library test version '0.0.000'

using QDM version '5.0.2'

valueset "Ethnicity": 'urn:oid:2.16.840.1.114222.4.11.837'

context Patient

define "SDE Ethnicity": ["Patient Characteristic Ethnicity": "Ethnicity"]

define "def": ["Encounter, Performed": "SDE Ethnicity"]
```

The following is the outputted ELM: 
```xml
<?xml version="1.0" encoding="UTF-8"?>
<library xmlns="urn:hl7-org:elm:r1" xmlns:t="urn:hl7-org:elm-types:r1" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:fhir="http://hl7.org/fhir" xmlns:a="urn:hl7-org:cql-annotations:r1">
   <identifier id="test" version="0.0.000"/>
   <schemaIdentifier id="urn:hl7-org:elm" version="r1"/>
   <usings>
      <def localIdentifier="System" uri="urn:hl7-org:elm-types:r1"/>
      <def localIdentifier="QDM" uri="urn:healthit-gov:qdm:v5_0_1_draft" version="5.0.2"/>
   </usings>
   <valueSets>
      <def name="Ethnicity" id="urn:oid:2.16.840.1.114222.4.11.837" accessLevel="Public"/>
   </valueSets>
   <statements>
      <def name="Patient" context="Patient">
         <expression xsi:type="SingletonFrom">
            <operand xmlns:ns0="urn:healthit-gov:qdm:v5_0_1_draft" dataType="ns0:Patient" templateId="Patient" xsi:type="Retrieve"/>
         </expression>
      </def>
      <def name="SDE Ethnicity" context="Patient" accessLevel="Public">
         <expression xmlns:ns1="urn:healthit-gov:qdm:v5_0_1_draft" dataType="ns1:PatientCharacteristicEthnicity" xsi:type="Retrieve">
            <codes name="Ethnicity" xsi:type="ValueSetRef"/>
         </expression>
      </def>
      <def name="def" context="Patient" accessLevel="Public">
         <expression xmlns:ns2="urn:healthit-gov:qdm:v5_0_1_draft" dataType="ns2:EncounterPerformed" templateId="PositiveEncounterPerformed" codeProperty="code" xsi:type="Retrieve">
            <codes name="SDE Ethnicity" xsi:type="ExpressionRef"/>
         </expression>
      </def>
   </statements>
</library>
```

Thoughts?

**Comment by cmoesel** [comment]
I'm pretty sure that's wrong.  As best I can remember, only a ValueSetRef should be allowable in a Retrieve.  It's possible that a list of codes or concepts is also allowed -- but even then, your example above is clearly not a list of codes or concepts.  @brynrhodes?

**Comment by jackcmeyer** [comment]
Doing a little more research into this...

When using the following cql:
```
library test version '0.0.000'

using QDM version '5.0.2'

valueset "Ethnicity": 'urn:oid:2.16.840.1.114222.4.11.837'

context Patient

define "def": ["Encounter, Performed": ["Patient Characteristic Ethnicity": "Ethnicity"]]
```
The cql-to-elm parser fails, with the following errors: 
```
Translation failed due to errors:
[9:39, 9:39] Syntax error at [
[9:1, 9:89] Internal translator error.
```

When running the cql from the original issue in the v12 branch, we get the same xml as before 
```xml
<?xml version="1.0" encoding="UTF-8"?>
<library xmlns="urn:hl7-org:elm:r1" xmlns:t="urn:hl7-org:elm-types:r1" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:fhir="http://hl7.org/fhir" xmlns:a="urn:hl7-org:cql-annotations:r1">
   <identifier id="test" version="0.0.000"/>
   <schemaIdentifier id="urn:hl7-org:elm" version="r1"/>
   <usings>
      <def localIdentifier="System" uri="urn:hl7-org:elm-types:r1"/>
      <def localIdentifier="QDM" uri="urn:healthit-gov:qdm:v5_0_1_draft" version="5.0.2"/>
   </usings>
   <valueSets>
      <def name="Ethnicity" id="urn:oid:2.16.840.1.114222.4.11.837" accessLevel="Public"/>
   </valueSets>
   <statements>
      <def name="Patient" context="Patient">
         <expression xsi:type="SingletonFrom">
            <operand xmlns:ns0="urn:healthit-gov:qdm:v5_0_1_draft" dataType="ns0:Patient" templateId="Patient" xsi:type="Retrieve"/>
         </expression>
      </def>
      <def name="SDE Ethnicity" context="Patient" accessLevel="Public">
         <expression xmlns:ns1="urn:healthit-gov:qdm:v5_0_1_draft" dataType="ns1:PatientCharacteristicEthnicity" xsi:type="Retrieve">
            <codes name="Ethnicity" xsi:type="ValueSetRef"/>
         </expression>
      </def>
      <def name="def" context="Patient" accessLevel="Public">
         <expression xmlns:ns2="urn:healthit-gov:qdm:v5_0_1_draft" dataType="ns2:EncounterPerformed" templateId="PositiveEncounterPerformed" codeProperty="code" xsi:type="Retrieve">
            <codes name="SDE Ethnicity" xsi:type="ExpressionRef"/>
         </expression>
      </def>
   </statements>
</library>
```

But, when running the cql from this comment, the cql-to-elm on the v12 branch parser passes validation and exports an ELM . 

```xml
<?xml version="1.0" encoding="UTF-8"?>
<library xmlns="urn:hl7-org:elm:r1" xmlns:t="urn:hl7-org:elm-types:r1" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:fhir="http://hl7.org/fhir" xmlns:a="urn:hl7-org:cql-annotations:r1">
   <identifier id="test" version="0.0.000"/>
   <schemaIdentifier id="urn:hl7-org:elm" version="r1"/>
   <usings>
      <def localIdentifier="System" uri="urn:hl7-org:elm-types:r1"/>
      <def localIdentifier="QDM" uri="urn:healthit-gov:qdm:v5_0_1_draft" version="5.0.2"/>
   </usings>
   <valueSets>
      <def name="Ethnicity" id="urn:oid:2.16.840.1.114222.4.11.837" accessLevel="Public"/>
   </valueSets>
   <statements>
      <def name="Patient" context="Patient">
         <expression xsi:type="SingletonFrom">
            <operand xmlns:ns0="urn:healthit-gov:qdm:v5_0_1_draft" dataType="ns0:Patient" templateId="Patient" xsi:type="Retrieve"/>
         </expression>
      </def>
      <def name="def" context="Patient" accessLevel="Public">
         <expression xmlns:ns1="urn:healthit-gov:qdm:v5_0_1_draft" dataType="ns1:EncounterPerformed" templateId="PositiveEncounterPerformed" codeProperty="code" xsi:type="Retrieve">
            <codes xmlns:ns2="urn:healthit-gov:qdm:v5_0_1_draft" dataType="ns2:PatientCharacteristicEthnicity" xsi:type="Retrieve">
               <codes name="Ethnicity" xsi:type="ValueSetRef"/>
            </codes>
         </expression>
      </def>
   </statements>
</library>
```

**Comment by brynrhodes** [comment]
It's wrong in 1.1, @cmoesel is correct only identifiers were syntactically allowed as the terminology target of a retrieve. However, the underlying ELM Retrieve has always allowed an arbitrary expression as the terminology target, and the syntactic restriction in CQL made some things unnecessarily difficult to express (consider asking for the components of a medication, using a GetIngredients() function that returns a list of codes, for example).

So this restriction was relaxed in CQL 1.2 and the terminology target of a retrieve is allowed to be an arbitrary expression. It must resolve to a list of codes, but it's no longer restricted to only identifiers.

**Comment by brynrhodes** [comment]
But you're right, it's not validating, that should fail semantically and doesn't. Fixing now.

**Comment by brynrhodes** [comment]
Fixed in 1.2.4-SNAPSHOT

**Comment by jackcmeyer** [comment]
@brynrhodes 

It looks like this issue is reappearing again. The following is my test file: 

```
library test version '0.0.000'

using QDM version '5.0.2'

valueset "Ethnicity": 'urn:oid:2.16.840.1.114222.4.11.837'

context Patient

define "SDE Ethnicity": ["Patient Characteristic Ethnicity": "Ethnicity"]

define "def": ["Encounter, Performed": "SDE Ethnicity"]
```

The ELM that is outputted is this: 
```xml
<?xml version="1.0" encoding="UTF-8"?>
<library xmlns="urn:hl7-org:elm:r1" xmlns:t="urn:hl7-org:elm-types:r1" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:fhir="http://hl7.org/fhir" xmlns:a="urn:hl7-org:cql-annotations:r1">
   <annotation startLine="11" startChar="15" endLine="11" endChar="55" message="Could not resolve membership operator for terminology target of the retrieve." errorType="semantic" errorSeverity="warning" xsi:type="a:CqlToElmError"/>
   <identifier id="test" version="0.0.000"/>
   <schemaIdentifier id="urn:hl7-org:elm" version="r1"/>
   <usings>
      <def localIdentifier="System" uri="urn:hl7-org:elm-types:r1"/>
      <def localIdentifier="QDM" uri="urn:healthit-gov:qdm:v5_0_1_draft" version="5.0.2"/>
   </usings>
   <valueSets>
      <def name="Ethnicity" id="urn:oid:2.16.840.1.114222.4.11.837" accessLevel="Public"/>
   </valueSets>
   <statements>
      <def name="Patient" context="Patient">
         <expression xsi:type="SingletonFrom">
            <operand xmlns:ns0="urn:healthit-gov:qdm:v5_0_1_draft" dataType="ns0:Patient" templateId="Patient" xsi:type="Retrieve"/>
         </expression>
      </def>
      <def name="SDE Ethnicity" context="Patient" accessLevel="Public">
         <expression xmlns:ns1="urn:healthit-gov:qdm:v5_0_1_draft" dataType="ns1:PatientCharacteristicEthnicity" codeProperty="code" xsi:type="Retrieve">
            <codes name="Ethnicity" xsi:type="ValueSetRef"/>
         </expression>
      </def>
      <def name="def" context="Patient" accessLevel="Public">
         <expression xmlns:ns2="urn:healthit-gov:qdm:v5_0_1_draft" dataType="ns2:PositiveEncounterPerformed" templateId="PositiveEncounterPerformed" codeProperty="code" xsi:type="Retrieve">
            <codes name="SDE Ethnicity" xsi:type="ExpressionRef"/>
         </expression>
      </def>
   </statements>
</library>
```

Note, there is a warning message in the `def` expression which says `Could not resolve membership operator for terminology target of the retrieve.`
